### ISSUE TITLE
initial thoughts regarding splitting up work for HackWars MMORPG

### ISSUE THREAD

**Issue by bcoe** [HackWars/docs#5]
Wanted to write down some initial thoughts I have about how I'd split up development of this game:
1. I'd create a shared module for the transport layer (how the world-map, compiler, etc, talk to the server components). I think socket.io would be a good technology to use for this:  
   - https://www.npmjs.com/package/socket.io-client
2. a codebase for the front-end world map:
   - https://github.com/HackWars/hackwars-world
3. a catch-all codebase for other front-end stuff:
   - your computer.
   - your character sheet.
   - your stats.
4. a backend server for the MMO.
5. an isolated server for executing user scripts:
   - https://github.com/bcoe/sandcastle

**Comment by jdog141176** [comment]
sadly i cant try out the "demo" you have out, i get errors thrown at me when i try to do npm install :/

**Comment by bcoe** [comment]
@jdog141176 mind sharing your `npm-debug.log` with me?

**Comment by jdog141176** [comment]
@bcoe 
0 info it worked if it ends with ok
1 verbose cli [ 'C:\Program Files\nodejs\\node.exe',
1 verbose cli   'C:\Program Files\nodejs\node_modules\npm\bin\npm-cli.js',
1 verbose cli   'install' ]
2 info using npm@2.11.3
3 info using node@v0.12.7
4 verbose config Skipping project config: C:\Users\joey/.npmrc. (matches userconfig)
5 verbose readDependencies loading dependencies from C:\Users\joey\package.json
6 error install Couldn't read dependencies
7 verbose stack Error: ENOENT, open 'C:\Users\joey\package.json'
7 verbose stack     at Error (native)
8 verbose cwd C:\Users\joey
9 error Windows_NT 6.3.9600
10 error argv "C:\Program Files\nodejs\\node.exe" "C:\Program Files\nodejs\node_modules\npm\bin\npm-cli.js" "install"
11 error node v0.12.7
12 error npm  v2.11.3
13 error path C:\Users\joey\package.json
14 error code ENOPACKAGEJSON
15 error errno -4058
16 error package.json ENOENT, open 'C:\Users\joey\package.json'
16 error package.json This is most likely not a problem with npm itself.
16 error package.json npm can't find a package.json file in your current directory.
17 verbose exit [ -4058, true ]

**Comment by jdog141176** [comment]
i know what i forgot... i forgot to clone the repo and run it from there haha

**Comment by jonpchin** [comment]
Is it possible you can package the javascript files into a .jsproj(javascript project file) so I can open it up on Microsoft Visual Community 2015? I am not sure how to use sandcastle. I tried going through your readme files on sandcastle but I am not certain if I should be just using solely the terminal or does sandcastle provide a GUI? I tried using Netbeans but it doesn't seem to have any way to manage javascript projects. Netbeans can only run single JS files.

**Comment by jonpchin** [comment]
I installed Node.js plugin in Visual Studio. I managed to get the JS project plugin working. Which file is the main startup file? I try running the main.js and it says it cannot find module pixi.js, I also tried running other files like hello.js and it says lodash is missing. I see lodash in another folder. 

Just looking at the code I don't see any import statements. I assume the npm somehow configures all the imports but I don't know how to set npm to do this for me. Any tips appreciated. 

Update: I finally got the npm install to work. There was an issue with the environment variables. I just got a new laptop and I never used Windows 8.1 before. And I don't have any experience in Node.js I am not sure what I am suppose to do next to start running the scripts. I'll do more research before I make any posts.

**Comment by BWHale1000** [comment]
It all depends on your development team. I know most of the developers for the original HackWars have a wide knowledge base so it depends on if you want to focus on one thing at a time and make sure it is easily compatible and implementable with your future work, or you could work on everything bit by bit and make sure it all works together. In my opinion(these are flying EVERYWHERE) the first way seems more idealistic because if you develop everything together you have a niche of a development team and the way things work together within the process of development, making it hard to work on in the future for anyone else but the original developers. If you developed it in the principle of a puzzle, fitting individual pieces together, you could potentially add/change things without worrying about the core of the game. I am not a developer myself so it could very well be nonsense or unrealistic but that's what opinions are for xD

**Comment by jonpchin** [comment]
I agree that we should plan this out carefully.  When I started toying around with sandcastle and trying to get npm and everything to work I was so confused on what to do. 

I want to help contribute.

**Comment by jdog141176** [comment]
i'm resigning from this, i can't even get the "demo" to work, if i can't even do that then i won't even be able to help,

also y'all should check out https://www.heroku.com/

**Comment by jonpchin** [comment]
Don't give up. Its normal to not understand how a specific technology works. Currently the documentation on how sandcastle seems lacking. Everything seems lacking and not newbie friendly. It could be I am just a newbie at javascript which I sort am. And yeah I saw that cloud hosting website a few weeks ago. There are also many other free cloud hosting alternatives popping up now.
